core/citrus-base/src/main/java/com/consol/citrus/util/SocketUtils.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2006-2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.consol.citrus.util;
+
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.util.Random;
+import javax.net.ServerSocketFactory;
+
+import com.consol.citrus.exceptions.CitrusRuntimeException;
+
+/**
+ * Find an available TCP port randomly selected from given port range.
+ */
+public class SocketUtils {
+
+    /**
+     * The minimum value for port ranges used when finding an available TCP port.
+     */
+    private static final int PORT_RANGE_MIN = 1024;
+
+    /**
+     * The maximum value for port ranges used when finding an available TCP port.
+     */
+    private static final int PORT_RANGE_MAX = 65535;
+
+    private static final int MAX_ATTEMPTS = 1000;
+
+    private static final Random random = new Random(System.nanoTime());
+
+
+    public static int findAvailableTcpPort() {
+        return findAvailableTcpPort(PORT_RANGE_MIN);
+    }
+
+    public static int findAvailableTcpPort(int min) {
+        if (min < PORT_RANGE_MIN) {
+            throw new CitrusRuntimeException(String.format("Invalid minimum value for port range - ports range must be greater than %d", PORT_RANGE_MIN));
+        }
+
+        int candidatePort;
+        int count = 0;
+        do {
+            if (++count > MAX_ATTEMPTS) {
+                throw new CitrusRuntimeException(String.format("Could not find an available TCP port in the range [%d, %d] after %d attempts", min, PORT_RANGE_MAX, MAX_ATTEMPTS));
+            }
+            candidatePort = min + random.nextInt(PORT_RANGE_MAX - min + 1);
+        }
+        while (!isPortAvailable(candidatePort));
+
+        return candidatePort;
+    }
+
+
+    /**
+     * Check if the specified TCP port is currently available on localhost.
+     */
+    private static boolean isPortAvailable(int port) {
+        try {
+            ServerSocket serverSocket = ServerSocketFactory.getDefault()
+                    .createServerSocket(port, 1, InetAddress.getByName("localhost"));
+            serverSocket.close();
```

Comment:
well, there is a reason spring removed it's implementation of this.. it's potentially unsafe and a source for flaky tests. why wouldn't you rely on the underlying systems to choose their port? I think it worked on the predecessor branch.. or didn't it?

Reply:
I think the reason that Spring has removed this utility is because it has been part of their production code base. Citrus is a test framework that usually is only used in test scope. Also Citrus itself is using this utility only in unit testing and in EmbeddedKafkaServer that also is only used in unit testing scope.

Given this context I still think that this utility is a clever and nice solution for its purpose in Citrus code base. Of course we can also bring back the solution from the predecessor branch in a follow up PR. I honestly do not remember why I have not used it in this PR - probably the merging effort has been a bit confusing for me in this area.